File: src/native/main/core/engine/dungeonutility/structures/StructureObject.cpp
#include <core/DungeonUtility.h>
#include <java.h>
#include <innercore_callbacks.h>
#include <api/BlockSource.h>

void DungeonUtility::StructureObject::addBlock(BlockData* block){
    this->blocks.push_back(block);
}

void DungeonUtility::StructureObject::set(int const x, int const y, int const z, BlockSource* region) const {
    auto it = blocks.begin();
    while (it != blocks.end()){
        (*it)->setBlock(x, y, z, region);
        it++;
    }
}

export(jlong, dungeonutility_struct_StructureObject_newStructureObject){
    return (jlong) new DungeonUtility::StructureObject();
}

export(void, dungeonutility_struct_StructureObject_addBlock, jlong pointer, jlong block){
    ((DungeonUtility::StructureObject*) pointer)->addBlock((DungeonUtility::BlockData*) block);
}

export(void, dungeonutility_struct_StructureObject_set, jlong pointer, jint x, jint y, jint z, jlong region){
    ((DungeonUtility::StructureObject*) pointer)->set((int) x, (int) y, (int) z, (BlockSource*) region);
}

File: src/java/Boot/src/com/core/api/dungeonutility/struct/StructureObject.java
package com.core.api.dungeonutility.struct;

import com.core.api.dungeonutility.api.BlockData;
import com.core.api.engine.PointerClass;
import com.zhekasmirnov.apparatus.mcpe.NativeBlockSource;

public class StructureObject extends PointerClass {
    native public static long newStructureObject();
    native public static void addBlock(long pointer, long block);
    native public static void set(long pointer, int x, int y, int z, long region);
    native public static long[] getBlocks(long pointer);

    public StructureObject(){
        super(newStructureObject());
    }

    public BlockData[] getBlocks(){
        long[] blocks = getBlocks(pointer);
        BlockData[] result = new BlockData[blocks.length];
        for(int i = 0;i < blocks.length;i++)
            result[i] = new BlockData(blocks[i]);
        return result;
    }

    public StructureObject(long pointer){
        super(pointer);
    }

    public StructureObject addBlock(BlockData block){
        addBlock(pointer, block.getPointer());
        return this;
    }

    public StructureObject set(int x, int y, int z, NativeBlockSource region){
        set(pointer, x, y, z, region.getPointer());
        return this;
    }
}
